### DISCUSSION TITLE
UAC prompt when connect to a client

### DISCUSSION THREAD

**Discussion by Sven1403** [rustdesk/rustdesk#6345] [General]
### Bug Description

Hello,

on some connections there is appearing a UAC window at the moment you want to connect to the client. The user on the other side can normally accept the connection and im able to do everything like it should be. But its still weird that the UAC prompt is coming up. It also confuse the user on the other side when the UAC window and the RustDesk Connection window appearing at the same time.

You know what this could be?

Greetings

### How to Reproduce

Hard to reproduce. Sometimes its coming up, sometimes not. I tested it on my work laptop, computer and VM and there its not coming up. I saw it like 4-5 times from my last 8 connections with users in my company

### Expected Behavior

No UAC window should come up when you want to connect to someone

### Operating system(s) on local side and remote side

Windows 10 -> Windows 10

### RustDesk Version(s) on local side and remote side

1.2.3 -> 1.2.3

### Screenshots

![grafik](https://github.com/rustdesk/rustdesk/assets/77511450/76a1d1d2-79bc-40f7-9b0e-8f4102492c6e)

**Comment by DMSnz** [comment]
The most common reason I would think of is to make firewall changes for non admin users?
But that is purely a guess.

**Reply by Sven1403** [reply]
Hello :)

thanks for the idea but the UAC window is coming up or is not coming up on users with the same privileges

**Comment by ukdt7** [comment]
ENGLISH (Translated with ChatGPT):
Hey!

I believe you are experiencing the same problems as I am. When you click on "Show more details," you'll see that a *.bat script wants to be executed from the Temp directory.

Ultimately, it tries to copy the Runtimebroker.exe from System32 into the program directory. It seems that Rustdesk does not use the Runtimebroker from System32. For whatever reason. I think the developers can explain that best.

For example, I have tried to omit the tray icon by deleting the autostart entry "%PROGRAMDATA%\Microsoft\Windows\Start Menu\Programs\Startup\RustDesk Tray.lnk". The shortcut starts rustdesk.exe --tray.

If the tray is not started or is terminated, then there is no Rustdesk service, or it is removed. As a result, Rustdesk runs entirely without SYSTEM permissions. I personally find that best. However, this in turn creates different problems:

If there is no background process with system rights, you can no longer respond to UAC prompts, because SecureDesktop access is only possible with SYSTEM rights. In addition, the Runtimebroker issue arises. Sometimes during the request, it appears that the user SYSTEM wants to connect. This should not happen.

**You can bypass the problem you mentioned** by renaming the Runtimebroker-Rustdesk.exe in the program directory. Then it works without UAC query. For whatever reason. I would also be interested in why the developers implemented it this way.

Now, for example, there is the function that you can use your admin credentials to see the SecureDesktop. However, this is somehow illogically solved. Because you can only see the option if the service has been deleted and the tray icon has been terminated. And when I then try to enter the AdminCredentials via Rustdesk, I just get the error message that it is not necessary.

In my opinion, Rustdesk can be operated under Windows completely without system rights and only needs an interface so that the supporter can start the service on the other side with system rights, so that the SecureDesktop (UAC messages) becomes visible **when needed**. Otherwise, I see significant security problems if a remote support tool is waiting in the background with system rights to receive commands.

DEUTSCH: Ich glaube du hast genau die gleichen Probleme, wie ich. Wenn du auf "Weitere Details anzeigen" klickst, siehst du dass ein *.bat Script aus dem Temp-Verzeichnis heraus ausgeführt werden möchte.

Dort wird im Endeffekt versucht die Runtimebroker.exe aus System32 in das Programmverzeichnis zu kopieren. Scheinbar nutzt Rustdesk nicht den Runtimebroker aus System32. Warum auch immer. Das können uns denke ich mal am besten die Entwickler erklären.

Ich zum Beispiel habe versucht das Traysymbol wegzulassen, in dem ich den Autostarteintrag "%PROGRAMDATA%\Microsoft\Windows\Start Menu\Programs\Startup\RustDesk Tray.lnk" lösche. Die Verknüpfung startet rustdesk.exe --tray.

Wenn die Tray nicht gestartet bzw. beendet wird, dann existiert kein Rustdesk Dienst bzw. er wird entfernt. Dadurch läuft Rustdesk insgesamt ohne SYSTEM-Rechte. Das finde ich persönlich am besten. Dadurch wiederum entstehen aber unterschiedliche Probleme:

Wenn es keinen Hintergrundprozess mit Systemrechten gibt, kann man nicht mehr auf UAC Aufforderungen reagieren, weil SecureDesktop Zugriff nur mit SYSTEM-Rechten möglich ist. Außerdem entsteht die Runtimebroker-Problematik. Manchmal erscheint auch bei der Anfrage, dass der User SYSTEM sich verbinden möchte. Das darf nicht passieren. 

**Man kann dein genanntes Problem umgehen**, indem man die Runtimebroker-Rustdesk.exe im Programmverzeichnis umbenennt. Dann funktioniert es ohne UAC Abfrage. Warum auch immer. Auch das würde mich interessieren, warum die Entwickler es so umgesetzt haben.

Jetzt gibt es zum Beispiel die Funktion, dass man seine Admincredentials benutzen kann, um den SecureDesktop sehen zu können. Jedoch ist das irgendwie unlogisch gelöst. Denn die Option kann man nur sehen, wenn der Service gelöscht und das Traysymbol beendet wurde. Und wenn ich dann versuche die AdminCredentials über Rustdesk einzugeben, dann kommt nur die Fehlermeldung, dass es nicht erforderlich sei. 

Meiner Meinung nach kann Rustdesk unter Windows vollständig ohne Systemrechte betrieben werden und benötigt nur eine Schnittstelle, damit der Supporter den Dienst auf der Gegenseite mit Systemrechten starten kann, damit der SecureDesktop (UAC Meldungen) **bei Bedarf** sichtbar wird. Ansonsten sehe ich große Sicherheitsprobleme, wenn ein Remotesupport Tool, im Hintergrund mit Systemrechten darauf wartet, Befehle entgegen zu nehmen.

**Reply by Sven1403** [reply]
damn thanks for this detailled answer mate. I will try your bypass :) maybe the developer will say something too in here.

**Reply by ukdt7** [reply]
i hope so. @rustdesk please help us :-D

**Comment by rustdesk** [comment]
Can you translate this picture?

![image](https://github.com/rustdesk/rustdesk/assets/71636191/8a5ecc41-0e8a-4d9d-9f9a-44971da55bc0)

**Reply by Sven1403** [reply]
Its just a normal UAC window.

I found this screenshot in google which is the same window:
![grafik](https://github.com/rustdesk/rustdesk/assets/77511450/d3b925b7-63e9-40af-815b-992b7ec7813b)

Next time i see this window, i will click on more details. I guess this will show, what exactly RustDesk wants to run with cmd

**Comment by rustdesk** [comment]
@21pages follow up this issue

**Reply by Sven1403** [reply]
@21pages 
Here some additional informations:
At the connection RustDesk wants to run this .bat
![grafik](https://github.com/rustdesk/rustdesk/assets/77511450/245104ae-ff37-45a5-a22d-2337f3739463)

The copy command in the .bat file is triggering the UAC window. When you delete/comment out the command and make the .bat file read-only, the UAC windows doesnt appear anymore. Read-only is important because at every connection the .bat file will be recreated
![grafik](https://github.com/rustdesk/rustdesk/assets/77511450/2d14df40-7697-49c6-939c-52d61f7f8781)

This could be a workaround (additional to the workaround from @ukdt7 ) but will this disable any functionality of RustDesk?

I hope this is helping a bit :)

**Reply by 21pages** [reply]
Thanks, it helps a lot.

**Comment by rustdesk** [comment]
@fufesou remove this if no privacy mode. Mentioned many times.

**Comment by rustdesk** [comment]
@Sven1403 please test the latest build, we have put fix. https://github.com/rustdesk/rustdesk/pull/6361

**Reply by Sven1403** [reply]
@rustdesk tested it on 2 computer where UAC windows popped up and now its not coming up anymore :) so its fixed. Waiting for the full release of 1.2.4 before rolling it out on all clients.

Thanks for fast fix! Multi-Monitor Feature is pretty cool btw ;)
